public/utils.py

```diff
@@ -118,3 +120,140 @@ def has_feature(feature):
+    """
+    Given DOI and URL for idresolve service, output query url for SFX
+    callback is called 'redundant' because this code isn't using it
+
+    Args:
+        string URL for idresolve, DOI string
+
+    Returns:
+        full URL for the idresolve API
+    """
+    output = (
+        f'{bare_url}'
+        '?code=9344'
+        '&function=idresolve'
+        '&callback=redundant'
+        f'&id={doi}'
+    )
+    return output
+
+
+def mk_url(doi):
+    """
+    Given just DOI, output query url for SFX
+
+    Args:
+        string DOI
+
+    Returns:
+        full URL for idresolve API
+    """
+    # change to the alternative for local testing
+    bare_url = "https://www.lib.uchicago.edu/cgi-bin/idresolve"
+    # bare_url = "your_local_domain:8081"
+    return full_mk_url(bare_url, doi)
+
+
+def doi_lookup(doi):
+    """
+    Query the DOI resolver service, return SFX URL if DOI is valid,
+    otherwise return None
+
+    Args:
+        non-validated, imperfect string DOI
+
+    Returns:
+        JSON response from idresolve, in string form
+    """
+    url = mk_url(doi)
+    try:
+        response = requests.get(url)
+    # if the idresolve service is down, doi_lookup should fail silently
+    except OSError:
+        return None
+    if response.status_code % 400 < 100:
+        return None
+    else:
+        return response.text
+
+
+def get_clean_params(request):
+    """
+    Return parameters that have been passed to a POST request,
+    omitting the CSRF token
+
+    Args:
+        a POST request
+
+    Returns:
+        POST parameters, in the form of a dictionary
+    """
+    params = request.POST
+    clean_params = dict(
+        filter(lambda x: x[0] != 'csrfmiddlewaretoken', params.items())
+    )
+    return clean_params
+
+
+def get_first_param(request):
+    """
+    Given a request, return the value of the first query string
+    parameter, whatever the key happens to be called
+
+    Args: 
+        a POST request
+
+    Returns:
+        string: the value of the first post parameter
+    """
+    params = get_clean_params(request)
+    if params:
+        first_key = list(params.keys())[0]
+        first_value = params[first_key]
+        return first_value
+    else:
+        return None
+
+
+def switchboard_url(form_name):
+    """
+    Map the name of each search form to the base URL used for the
+    relevant search
+
+    Args: 
+        a string key indicating the type of form on the main page
+
+    Returns:
+        the URL to post to for the relevant search box
+    """
+    if form_name == 'catalog':
+        return 'https://catalog.lib.uchicago.edu/vufind/Search/Results'
+    elif form_name == 'articles':
+        url = (
+            'http://proxy.uchicago.edu/login'
+            '?url=http://search.ebscohost.com/login.aspx'
+            '?direct=true&site=eds-live'
+            '&scope=site'
+            '&type=0'
+            '&mode=and'
+            '&cli0=FT1&clv0=Y'
+        )
```

Comment:
I was unfamiliar with this way of doing a multi-line string in python. It seems to work fine in the interpreter. Cool!